job/job.go

```diff
@@ -0,0 +1,308 @@
+	// If job has already exited, then disallow rerun
+	// NOTE: user can always create a new job with same command
+	if j.status.Status == EXITED {
+		return errors.New("job has exited; cannot restart")
+	}
+	j.Unlock()
+
+	// If creating a file to store output fails, don't run the job
+	// TODO: this is not ideal solution. Ideally, we want to fall
+	// back to in-memory buffer for holding the output.
+	// and stream by reading the buffer.
+	// TODO: implement the this if have some time
+	outStreamer, err := NewStreamer()
+	if err != nil {
+		return errors.New("failed to create a file to stream output of the run")
+	}
+
+	j.outStreamer = outStreamer
+
+	// start() will call exec.Cmd.Start() which might return an error
+	// need to wait until result of Start() comes back so that we can
+	// immediately report it back to the caller
+	errC := make(chan error)
+	go j.start(errC)
+
+	// Channel will either return an error or will be closed and give nil error
+	// In any case, we need to return it back to the caller
+	err = <-errC
+
+	return err
+}
+
+// Status returns status of this job.
+// Users can determine whether specified job has completed or not via this function
+func (j *Job) Status() JobStatus {
+	j.Lock()
+	defer j.Unlock()
+	return j.status
+}
+
+// Stop terminates a running job. Returns immediately if job has already exited
+func (j *Job) Stop() error {
+	j.Lock()
+	if j.status.Status == EXITED {
+		return nil
+	}
+	j.Unlock()
+
+	pid, err := j.Pid()
+	if err != nil {
+		return err
+	}
+
+	pgid, err := syscall.Getpgid(pid)
+	if err != nil {
+		return err
+	}
+
+	err = syscall.Kill(-pgid, syscall.SIGTERM)
+	if err != nil {
+		return err
+	}
+
+	return err
+}
+
+// Output returns a string channel which will be closed by library internally
+// once the job completes. Multiple calls to this function will create separate channels
+// so it is safe to stream output to multiple clients.
+func (j *Job) Output() chan []byte {
+	j.Lock()
+	defer j.Unlock()
+	rd, err := j.outStreamer.NewReader()
+	if err != nil {
+		return nil
+	}
+
+	// create streaming channel and start pushing
+	stream := make(chan []byte)
+	go j.startStream(rd, stream)
+	return stream
+}
+
+// Wait blocks until Job is complete.
+func (j *Job) Wait() {
+	<-j.doneC
+}
+
+// Pid returns the PID of underlying process
+func (j *Job) Pid() (int, error) {
+	j.Lock()
+	defer j.Unlock()
+	if j.cmd.Process == nil {
+		return -1, errors.New("job has not started yet")
+	}
+	return j.cmd.Process.Pid, nil
+}
+
+func (j *Job) SetUsername(username string) {
+	j.Lock()
+	defer j.Unlock()
+
+	j.username = username
+}
+
+func (j *Job) Username() string {
+	j.Lock()
+	defer j.Unlock()
+
+	return j.username
+}
+
+// startStream takes a ReadCloser and reads data line by line
+// then pushes them into the provided channel until Job is exited
+func (j *Job) startStream(rd io.ReadCloser, stream chan []byte) {
+	defer close(stream)
+	brd := bufio.NewReader(rd)
+	for {
+		line, err := brd.ReadBytes('\n')
```

Comment:
"ReadBytes reads until the first occurrence of delim in the input, returning a slice containing the data up to and including the delimiter. If ReadBytes encounters an error before finding a delimiter, **it returns the data read before the error and the error itself (often io.EOF). ReadBytes returns err != nil if and only if the returned data does not end in delim.**"

If a command generates output in a single line without '\n' byte, then ReadBytes() will return whatever it can read and io.EOF. Inside the if statement below, I should check if line buffer size is non-zero, then simply write the byte slice to the stream channel. It should handle the case.